Configuration/PyReleaseValidation/python/relval_steps.py

```diff
@@ -2466,20 +2468,66 @@ def gen2018HiMix(fragment,howMuch):
 
     # setup PU
     if k2 in PUDataSets:
+        # Setup premixing stage1
+        #
+        # Has to be done before the overall PU definition in order to benefit from that
+        #
+        # It is a complete overkill to define premixing step for
+        # each generator fragment, but let's worry about
+        # simplification later
+        for step in upgradeSteps['baseline']['steps']:
+            if "GenSim" in step:
+                stepNamePmx = step.replace('GenSim', 'Premix') + 'PU' + upgradeSteps['Premix']['suffix']
+                d = merge([{'-s'            : 'GEN,SIM,DIGI:pdigi_valid,L1,DIGI2RAW',
+                            '--datatier'    : 'PREMIX',
+                            '--eventcontent': 'PREMIX',
+                            '--procModifiers': 'premix_stage1',
+                           },
+                           upgradeStepDict[stepName][k]])
```

Comment:
this needs to be `upgradeStepDict[stepNamePmx][k]]` or else you get `heCollapse` flags

Reply:
> this needs to be upgradeStepDict[stepNamePmx][k]]

Umm, no. At this stage `upgradeStepDict[stepNamePmx][k]` is empty (and is actually set on the next line). The point here is to take the baseline PU workflow, and replace these parameters for premixing.

> or else you get heCollapse flags

My understanding is that `merge` gives priority on the first elements of the argument list
https://github.com/cms-sw/cmssw/blob/45370be098d9bd1b2178e600ebf3dfaf92b34494/Configuration/PyReleaseValidation/python/MatrixUtil.py#L192-L193
so actually the `hecollapse` would be replaced with `premix_stage1`. 

But this raises a point that for some parameters (`procModifiers`, `customise`, maybe also `era`) sometimes one wants to concatenate the arguments rather than replace them. For `customise` there are already copy-pasted pattern of "check manually if it exists, then manually append, else set", but given that the need for appending starts to spread, maybe it would be good to think more "central" support for that in `merge`? I can also try to think of something.

Reply:
I think there are actually two problems:
1. The variable `stepName` is not defined anywhere in this loop, so it picks up the last definition, which is in the `heCollapse` modification. Indeed, this shouldn't be `stepNamePmx` (I looked too quickly), but needs to be something.
2. The `heCollapse` modifications use `--procModifier` instead of `--procModifiers`; this is a typo that should be fixed.

The idea about concatenating in merge is a good one, but separate from the problem I observed here.

Reply:
> 1. The variable stepName is not defined anywhere in this loop, so it picks up the last definition, which is in the heCollapse modification. Indeed, this shouldn't be stepNamePmx (I looked too quickly), but needs to be something.

Good catch (what was I thinking...). The `stepName` here should be `step + 'PU' + upgradeSteps[stepType]['suffix']`, but now I have to re-think why did I place this piece here instead of being below before the line 2497 (where the lines 2493-2494 set the `stepName` for this piece).

> 2. The heCollapse modifications use --procModifier instead of --procModifiers; this is a typo that should be fixed.

I agree, see #22437.

Reply:
@kpedro88 For 1, I added the definition of `stepName` before its use and did some minor related cleanup. The generated configuration looks fine now.

Reply:
I created an issue #22492 to remind the concatenation option to `merge()`.